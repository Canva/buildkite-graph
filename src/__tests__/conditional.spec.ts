import {
  CommandStep,
  Conditional,
  Pipeline,
  Step,
  ThingOrGenerator,
} from '../';
import { createTest, serializers } from './helpers';

class MyConditional<T extends Step> extends Conditional<T> {
  constructor(
    step: ThingOrGenerator<T>,
    private readonly accepted: ReturnType<Conditional<T>['accept']>,
  ) {
    super(step as any);
  }

  accept(): ReturnType<Conditional<T>['accept']> {
    return this.accepted;
  }
}

describe('buildkite-graph', () => {
  describe('Steps', () => {
    describe('Command', () => {
      createTest('step addition', () => [
        new Pipeline('whatever').add(
          new MyConditional(new CommandStep('yarn').add('yarn test'), true),
        ),
        new Pipeline('whatever').add(
          new MyConditional(
            () => new CommandStep('yarn').add('yarn test'),
            true,
          ),
        ),
        new Pipeline('whatever').add(
          new MyConditional(new CommandStep('yarn').add('yarn test'), false),
        ),
      ]);

      createTest('async step addition', () => [
        new Pipeline('whatever').add(
          new MyConditional(
            new CommandStep('yarn').add('yarn test'),
            Promise.resolve(true),
          ),
        ),
        new Pipeline('whatever').add(
          new MyConditional(
            new CommandStep('yarn').add('yarn test'),
            Promise.resolve(false),
          ),
        ),
      ]);

      createTest('async step creation', () => [
        new Pipeline('whatever').add(
          new MyConditional(
            Promise.resolve(new CommandStep('yarn').add('yarn test')),
            true,
          ),
        ),
        new Pipeline('whatever').add(
          new MyConditional(
            () => Promise.resolve(new CommandStep('yarn').add('yarn test')),
            true,
          ),
        ),
      ]);

      it('throws on accept rejection', async () => {
        expect(
          serializers.json.serialize(
            new Pipeline('whatever').add(
              new MyConditional(
                new CommandStep('yarn').add('yarn test'),
                Promise.reject(new Error('O noes!!!')),
              ),
            ),
          ),
        ).rejects.toThrowError();
      });

      describe('Conditional dependencies', () => {
        createTest('can be specified', () => {
          const p = new Pipeline('x');

          // even though the onditional is set to false,
          // "a" will be added to the graph as "b" depends on it
          const a = new MyConditional(new CommandStep('a'), false);
          p.add(new CommandStep('b').dependsOn(a));

          return p;
        });

        describe('can be specified multiple times', () => {
          createTest('as dependency', () => {
            const p = new Pipeline('x');

            // even though the onditional is set to false,
            // "a" will be added to the graph as "b" depends on it
            const a = new MyConditional(new CommandStep('a'), false);
            p.add(new CommandStep('b').dependsOn(a));
            p.add(new CommandStep('c').dependsOn(a));

            return p;
          });
          it('but not in the pipeline', () => {
            expect(() => {
              const a = new MyConditional(new CommandStep('a'), false);
              new Pipeline('x').add(a, a);
            }).toThrow();
          });
        });

        it('conditionals are only unwrapped once', () => {
          const p = new Pipeline('x');

          const gen = jest.fn();
          gen.mockReturnValueOnce(new CommandStep('a'));
          gen.mockImplementation(() => {
            throw new Error('only once!');
          });
          const a = new MyConditional(gen, false);
          p.add(new CommandStep('b').dependsOn(a));

          serializers.json.serialize(p);
        });
      });

      describe.only('isEffectOf', () => {
        createTest(
          'will add steps if their effect dependency is accepted',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests'),
              true,
            );
            const deployCoverage = new CommandStep(
              'deploy coverage',
            ).isEffectOf(acceptedTests);

            return new Pipeline('x').add(acceptedTests, deployCoverage);
          },
          ['structure'],
        );

        createTest(
          'will not add steps if effect dependency is rejected',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests'),
              false,
            );
            const deployCoverage = new CommandStep(
              'deploy coverage',
            ).isEffectOf(acceptedTests);

            return new Pipeline('x').add(acceptedTests, deployCoverage);
          },
          ['structure'],
        );

        createTest(
          'will add steps if acceptAllConditions is set even if effect dependency is rejected',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests'),
              false,
            );
            const deployCoverage = new CommandStep(
              'deploy coverage',
            ).isEffectOf(acceptedTests);

            return new Pipeline('x').add(acceptedTests, deployCoverage);
          },
          ['json_depends_on_accept_all', 'yaml_depends_on_accept_all'],
        );

        createTest(
          'will not add steps if any effect dependency is rejected',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests 1'),
              true,
            );
            const rejectedTests = new MyConditional(
              new CommandStep('run tests 2'),
              false,
            );
            const deployCoverage = new CommandStep(
              'deploy coverage',
            ).isEffectOf(rejectedTests, acceptedTests);

            return new Pipeline('x').add(
              acceptedTests,
              rejectedTests,
              deployCoverage,
            );
          },
          ['structure'],
        );

        createTest(
          'effects of effects will be added if first effect dependency is accepted',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests'),
              true,
            );
            const createCoverageStep = new CommandStep(
              'create coverage',
            ).isEffectOf(acceptedTests);
            const deployCoverage = new CommandStep(
              'deploy coverage',
            ).isEffectOf(createCoverageStep);

            return new Pipeline('x').add(
              acceptedTests,
              createCoverageStep,
              deployCoverage,
            );
          },
          ['structure'],
        );

        createTest(
          'effects of effects will not be added if first effect dependency is rejected',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests'),
              false,
            );
            const createCoverageStep = new CommandStep(
              'create coverage',
            ).isEffectOf(acceptedTests);
            const deployCoverage = new CommandStep(
              'deploy coverage',
            ).isEffectOf(createCoverageStep);

            return new Pipeline('x').add(
              acceptedTests,
              createCoverageStep,
              deployCoverage,
            );
          },
          ['structure'],
        );

        createTest(
          'last call wins',
          () => {
            const acceptedTests = new MyConditional(
              new CommandStep('run tests'),
              false,
            );
            const deployCoverage1 = new CommandStep('deploy coverage')
              .isEffectOf(acceptedTests)
              .dependsOn(acceptedTests);
            const deployCoverage2 = new CommandStep('deploy coverage')
              .dependsOn(acceptedTests)
              .isEffectOf(acceptedTests);

            return [
              new Pipeline('x').add(deployCoverage1),
              new Pipeline('x').add(deployCoverage2),
            ];
          },
          ['structure'],
        );

        createTest(
          'dependsOn is used after isEffectOf',
          () => {
            const buildConditional = new MyConditional(
              new CommandStep('build app'),
              false,
            );
            const tests = new MyConditional(
              new CommandStep('run tests'),
              false,
            );

            const deployApp = new CommandStep('deploy app')
              .isEffectOf(buildConditional)
              .dependsOn(tests);

            return new Pipeline('x').add(buildConditional, deployApp);
          },
          ['structure'],
        );

        createTest(
          'effects of steps that are becoming part of the graph are exercised',
          () => {
            const buildConditional = new MyConditional(
              new CommandStep('build app'),
              false,
            );
            const tests = new MyConditional(
              () =>
                new CommandStep('run integration tests').dependsOn(
                  buildConditional,
                ),
              true,
            );

            const deployApp = new CommandStep('deploy app').isEffectOf(
              buildConditional,
            );

            return new Pipeline('x').add(buildConditional, tests, deployApp);
          },
          ['structure'],
        );

        createTest(
          'later steps affect earlier effects',
          () => {
            const p = new Pipeline('x');
            const buildConditional = new MyConditional(
              () => new CommandStep('build app'),
              false,
            );
            p.add(buildConditional);

            const deployApp = new CommandStep('deploy app').isEffectOf(
              buildConditional,
            );
            p.add(deployApp);

            const releaseApp = new CommandStep('release app').isEffectOf(
              deployApp,
            );
            p.add(releaseApp);

            const tests = new MyConditional(
              () =>
                new Promise<Step>((resolve) =>
                  setTimeout(
                    () =>
                      resolve(
                        new CommandStep('run integration tests').dependsOn(
                          buildConditional,
                        ),
                      ),
                    100,
                  ),
                ),
              true,
            );
            p.add(tests);

            return p;
          },
          ['structure'],
        );

        createTest(
          'effects and conditionals have correct depends_on',
          () => {
            const p = new Pipeline('x');
            const buildConditional = new MyConditional(
              () => new CommandStep('build app').withKey('build'),
              false,
            );
            p.add(buildConditional);

            const deployStep = new CommandStep('deploy app')
              .withKey('deploy')
              .isEffectOf(buildConditional);
            p.add(deployStep);

            const testsUsingBuildArtifact = new CommandStep('ssr tests')
              .withKey('ssr-tests')
              .dependsOn(buildConditional);
            p.add(testsUsingBuildArtifact);
            return p;
          },
          ['yaml_depends_on'],
        );
      });
    });
  });
});
