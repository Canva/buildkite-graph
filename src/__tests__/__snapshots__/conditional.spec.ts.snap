// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildkite-graph Steps Command Conditional dependencies can be specified dot 1`] = `
"digraph \\"x\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<a>\\" [ color = \\"grey\\" ];
}

subgraph cluster_1 {
  graph [ color = \\"black\\" ];
  \\"<b>\\" [ color = \\"grey\\" ];
}

  \\"[object Object]\\";
  \\"<b>\\";
  \\"[object Object]\\" -> \\"<b>\\" [ ltail = \\"cluster_0\\", lhead = \\"cluster_1\\" ];
}
"
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified json 1`] = `
Object {
  "steps": Array [
    Object {
      "command": "a",
    },
    Object {
      "wait": null,
    },
    Object {
      "command": "b",
    },
  ],
}
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified multiple times as dependency dot 1`] = `
"digraph \\"x\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<a>\\" [ color = \\"grey\\" ];
}

subgraph cluster_1 {
  graph [ color = \\"black\\" ];
  \\"<b>\\" [ color = \\"grey\\" ];
  \\"<c>\\" [ color = \\"grey\\" ];
}

  \\"[object Object]\\";
  \\"<b>\\";
  \\"<c>\\";
  \\"[object Object]\\" -> \\"<b>\\" [ ltail = \\"cluster_0\\", lhead = \\"cluster_1\\" ];
  \\"[object Object]\\" -> \\"<c>\\" [ ltail = \\"cluster_0\\", lhead = \\"cluster_1\\" ];
}
"
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified multiple times as dependency json 1`] = `
Object {
  "steps": Array [
    Object {
      "command": "a",
    },
    Object {
      "wait": null,
    },
    Object {
      "command": "b",
    },
    Object {
      "command": "c",
    },
  ],
}
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified multiple times as dependency structure 1`] = `
"* <a>
* [wait]
* <b>
* <c>"
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified multiple times as dependency yaml 1`] = `
"steps:
  - command: a
  - wait: ~
  - command: b
  - command: c
"
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified structure 1`] = `
"* <a>
* [wait]
* <b>"
`;

exports[`buildkite-graph Steps Command Conditional dependencies can be specified yaml 1`] = `
"steps:
  - command: a
  - wait: ~
  - command: b
"
`;

exports[`buildkite-graph Steps Command async step addition dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<yarn && yarn test>\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Command async step addition dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
}
"
`;

exports[`buildkite-graph Steps Command async step addition json 1`] = `
Object {
  "steps": Array [
    Object {
      "command": Array [
        "yarn",
        "yarn test",
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Command async step addition json 2`] = `
Object {
  "steps": Array [],
}
`;

exports[`buildkite-graph Steps Command async step addition structure 1`] = `"* <yarn && yarn test>"`;

exports[`buildkite-graph Steps Command async step addition structure 2`] = `""`;

exports[`buildkite-graph Steps Command async step addition yaml 1`] = `
"steps:
  - command:
      - yarn
      - yarn test
"
`;

exports[`buildkite-graph Steps Command async step addition yaml 2`] = `
"steps: []
"
`;

exports[`buildkite-graph Steps Command async step creation dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<yarn && yarn test>\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Command async step creation dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<yarn && yarn test>\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Command async step creation json 1`] = `
Object {
  "steps": Array [
    Object {
      "command": Array [
        "yarn",
        "yarn test",
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Command async step creation json 2`] = `
Object {
  "steps": Array [
    Object {
      "command": Array [
        "yarn",
        "yarn test",
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Command async step creation structure 1`] = `"* <yarn && yarn test>"`;

exports[`buildkite-graph Steps Command async step creation structure 2`] = `"* <yarn && yarn test>"`;

exports[`buildkite-graph Steps Command async step creation yaml 1`] = `
"steps:
  - command:
      - yarn
      - yarn test
"
`;

exports[`buildkite-graph Steps Command async step creation yaml 2`] = `
"steps:
  - command:
      - yarn
      - yarn test
"
`;

exports[`buildkite-graph Steps Command isEffectOf dependsOn is used after isEffectOf structure 1`] = `""`;

exports[`buildkite-graph Steps Command isEffectOf effects of effects will be added if first effect dependency is accepted structure 1`] = `
"* <run tests>
* [wait]
* <create coverage>
* [wait]
* <deploy coverage>"
`;

exports[`buildkite-graph Steps Command isEffectOf effects of effects will not be added if first effect dependency is rejected structure 1`] = `""`;

exports[`buildkite-graph Steps Command isEffectOf last call wins structure 1`] = `
"* <run tests>
* [wait]
* <deploy coverage>"
`;

exports[`buildkite-graph Steps Command isEffectOf last call wins structure 2`] = `""`;

exports[`buildkite-graph Steps Command isEffectOf will add steps if their effect dependency is accepted structure 1`] = `
"* <run tests>
* [wait]
* <deploy coverage>"
`;

exports[`buildkite-graph Steps Command isEffectOf will not add steps if any effect dependency is rejected structure 1`] = `"* <run tests 1>"`;

exports[`buildkite-graph Steps Command isEffectOf will not add steps if effect dependency is rejected structure 1`] = `""`;

exports[`buildkite-graph Steps Command step addition dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<yarn && yarn test>\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Command step addition dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"<yarn && yarn test>\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Command step addition dot 3`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
}
"
`;

exports[`buildkite-graph Steps Command step addition json 1`] = `
Object {
  "steps": Array [
    Object {
      "command": Array [
        "yarn",
        "yarn test",
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Command step addition json 2`] = `
Object {
  "steps": Array [
    Object {
      "command": Array [
        "yarn",
        "yarn test",
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Command step addition json 3`] = `
Object {
  "steps": Array [],
}
`;

exports[`buildkite-graph Steps Command step addition structure 1`] = `"* <yarn && yarn test>"`;

exports[`buildkite-graph Steps Command step addition structure 2`] = `"* <yarn && yarn test>"`;

exports[`buildkite-graph Steps Command step addition structure 3`] = `""`;

exports[`buildkite-graph Steps Command step addition yaml 1`] = `
"steps:
  - command:
      - yarn
      - yarn test
"
`;

exports[`buildkite-graph Steps Command step addition yaml 2`] = `
"steps:
  - command:
      - yarn
      - yarn test
"
`;

exports[`buildkite-graph Steps Command step addition yaml 3`] = `
"steps: []
"
`;
